### PR TITLE
[move-vm][aptos-vm] Basic execution counters

### PR DESCRIPTION
**Pull request by georgemitenkov**
## Description

Some basic counters that address #14348. Note that some metrics should not be collected via this timer, e.g., interpreter loop so that there is no perf degradation. Hence, added only in a few interesting places (e.g., module loading cache misses, resource loading cache misses, total execution times for user transaction with basic break downs).

## Type of Change
- [ ] New feature
- [ ] Bug fix
- [ ] Breaking change
- [ ] Performance improvement
- [ ] Refactoring
- [ ] Dependency update
- [ ] Documentation update
- [ ] Tests

## Which Components or Systems Does This Change Impact?
- [ ] Validator Node
- [ ] Full Node (API, Indexer, etc.)
- [ ] Move/Aptos Virtual Machine
- [ ] Aptos Framework
- [ ] Aptos CLI/SDK
- [ ] Developer Infrastructure
- [ ] Move Compiler
- [ ] Other (specify)

## Checklist
- [ ] I have read and followed the [CONTRIBUTING](https://github.com/aptos-labs/aptos-core/blob/main/CONTRIBUTING.md) doc
- [ ] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I identified and added all stakeholders and component owners affected by this change as reviewers
- [ ] I tested both happy and unhappy path of the functionality
- [ ] I have made corresponding changes to the documentation

### CODE REVIEW

**Comment by georgemitenkov** [comment]
* **#15086** <a href="https://app.graphite.dev/github/pr/aptos-labs/aptos-core/15086?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈
* `main`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @georgemitenkov and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment">Graphite <b>Graphite</b></a>

**Review by brmataptos — APPROVED** [review]
Thanks!

File: third_party/move/move-vm/metrics/src/lib.rs [hunk 1]
```diff
@@ -0,0 +1,37 @@
+// Copyright (c) The Move Contributors
+// SPDX-License-Identifier: Apache-2.0
+
+use once_cell::sync::Lazy;
+use prometheus::{register_histogram_vec, HistogramTimer, HistogramVec};
+
+pub trait Timer {
+    fn timer_with_label(&self, label: &str) -> HistogramTimer;
```
Comment (vineethk) [inline third_party/move/move-vm/metrics/src/lib.rs:8]: Given this is a `pub trait`, could we add a brief doc string for this method? Also, some guidance regarding the naming of the label, and any notes on performance considerations.

File: third_party/move/move-vm/metrics/src/lib.rs [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+
+    register_histogram_vec!(
+        // Metric name:
+        "vm_timer_seconds",
+        // Metric description:
+        "VM timers",
+        &["name"],
+        buckets,
+    )
+    .unwrap()
```
Comment (vineethk) [inline third_party/move/move-vm/metrics/src/lib.rs:36]: `expect()` with a message instead?

File: third_party/move/move-vm/metrics/src/lib.rs [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+
+pub static VM_TIMER: Lazy<HistogramVec> = Lazy::new(|| {
+    let factor = 2.0;
+    let num_buckets = 32;
+
+    let mut next = 1e-9;
+    let mut buckets = Vec::with_capacity(num_buckets);
+    for _ in 0..num_buckets {
+        buckets.push(next);
+        next *= factor;
```
Comment (vineethk) [inline third_party/move/move-vm/metrics/src/lib.rs:25]: Are we setting up a log graph here? Might be worth saying that in a doc string for `VM_TIMER`.

File: third_party/move/move-vm/metrics/src/lib.rs [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+    fn timer_with_label(&self, label: &str) -> HistogramTimer;
+}
+
+impl Timer for HistogramVec {
+    fn timer_with_label(&self, label: &str) -> HistogramTimer {
+        self.with_label_values(&[label]).start_timer()
+    }
+}
+
+pub static VM_TIMER: Lazy<HistogramVec> = Lazy::new(|| {
```
Comment (vineethk) [inline third_party/move/move-vm/metrics/src/lib.rs:17]: Was not clear to me where the results are dumped/can be accessed? Can we see some results showing all this in action for some transactions, demonstrating that we are getting sane results?

**Review by vineethk — APPROVED** [review]
Approving with comments to be addressed.

File: third_party/move/move-vm/metrics/src/lib.rs (hunk 4 above)
Comment (georgemitenkov) [reply third_party/move/move-vm/metrics/src/lib.rs:17]: The results can be seen on grafana dashboard, I will add a comment for this👍

Comment (vineethk) [reply third_party/move/move-vm/metrics/src/lib.rs:17]: Ah, nice. Can we also see results locally, e.g., when locally replaying a transaction?

Comment (georgemitenkov) [reply third_party/move/move-vm/metrics/src/lib.rs:17]: Yes, if you set up some env variables you can push metrics into grafana, and view them real-time:
```
export PUSH_METRICS_NAMESPACE=...
export PUSH_METRICS_ENDPOINT="..."
```

File: third_party/move/move-vm/metrics/src/lib.rs (hunk 2 above)
Comment (georgemitenkov) [reply third_party/move/move-vm/metrics/src/lib.rs:36]: Done

File: third_party/move/move-vm/metrics/src/lib.rs (hunk 1 above)
Comment (georgemitenkov) [reply third_party/move/move-vm/metrics/src/lib.rs:8]: Good point, done

File: third_party/move/move-vm/metrics/src/lib.rs (hunk 4 above)
Comment (georgemitenkov) [reply third_party/move/move-vm/metrics/src/lib.rs:17]: Added a comment about this

File: third_party/move/move-vm/metrics/src/lib.rs (hunk 3 above)
Comment (georgemitenkov) [reply third_party/move/move-vm/metrics/src/lib.rs:25]: Added

File: third_party/move/move-vm/metrics/src/lib.rs [hunk 5]
```diff
@@ -0,0 +1,46 @@
+// Copyright (c) The Move Contributors
+// SPDX-License-Identifier: Apache-2.0
+
+use once_cell::sync::Lazy;
+use prometheus::{register_histogram_vec, HistogramTimer, HistogramVec};
+
+/// Helper trait to encapsulate [HistogramVec] functionality. Users can use this trait to time
+/// different VM parts collecting metrics for different labels. Use wisely as timers do introduce
+/// an overhead, so using on extremely hot path is not recommended.
```
Comment (igor-aptos) [inline third_party/move/move-vm/metrics/src/lib.rs:9]: I think that this is too optimistic of a wording, let's at least remove word "extremely".  Even a moderate hot paths can cause slowdown, as histogram is not that cheap. 

Things in this PR are probably fine (I assume type_to_type_tag is not called too often?), but it is easy do add overhead (i.e
 Cache hot for data would probably be questionable)

**Review by igor-aptos — APPROVED** [review]

File: third_party/move/move-vm/metrics/src/lib.rs (hunk 5 above)
Comment (georgemitenkov) [reply third_party/move/move-vm/metrics/src/lib.rs:9]: Good point! 

Was testing with e2e, data accesses are indeed something we do not want to record, as there is a bit of regression. At least recording gas traversal & loading/verification miss + total execution time should be enough

**Pull request merged**